src/NServiceBus.Core/Causation/AttachCausationHeadersBehavior.cs

```diff
@@ -44,11 +49,36 @@ static void SetConversationIdHeader(IOutgoingLogicalMessageContext context, Inco
 
             if (hasUserDefinedConversationId)
             {
-                // do not override user defined conversation id if no incoming message exists.
                 return;
             }
 
-            context.Headers[Headers.ConversationId] = CombGuid.Generate().ToString();
+            string conversationId;
+            bool userProvidedId;
+
+            try
+            {
+                userProvidedId = tryGetConversationIdDelegate(new ConversationIdStrategyContext(context.Message), out conversationId);
+            }
+            catch (Exception exception)
+            {
+                throw new Exception($"Failed to execute CustomConversationId delegate. This configuration option was defined using {nameof(EndpointConfiguration)}.{nameof(MessageCausationConfigurationExtensions.CustomConversationIdStrategy)}.", exception);
+            }
+
+            if (userProvidedId)
+            {
+                if (string.IsNullOrEmpty(conversationId))
+                {
+                    throw new Exception("Null or empty conversation ID's are not allowed.");
+                }
+            }
+            else
+            {
+                conversationId = CombGuid.Generate().ToString();
```

Comment:
Is it really that valuable to allow both overriding the id generation and still fall back to a default in some cases? It doesn't seem like too much of a burden to say "you've overridden this, so you now have to provide us one all the time."